Add support for invitations that never expire.

sahil839 opened the pull request:
<!-- What's this PR for?  (Just a link to an issue is fine.) -->
- First commit adds backend support for invitations that never expire. We save `expiry_date` as None in the database for such invitations and use `-1` as the API level value.
- Second commit adds frontend support for the same.

Related Issue - #20337
 <!-- How have you tested? -->

<!-- If a UI change.  See:
  https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
  -->

<!-- Also be sure to make clear, coherent commits:
  https://zulip.readthedocs.io/en/latest/contributing/version-control.html
  -->

sahil839 commented on confirmation/models.py:
```diff
@@ -100,8 +100,11 @@ def create_confirmation_link(
-    if validity_in_days:
-        expiry_date = current_time + datetime.timedelta(days=validity_in_days)
+    if validity_in_days is not None:
+        if validity_in_days == -1:
```
We currently use `-1` as API level value. We can also use a string like `never` but I feel using `-1` is better.

timabbott replied:
Probably a good question for "API design".

sahil839 commented:
Do we want to send reminder invitation emails as well for these invites which would never expire? We currently send reminder emails 2 days before the expiry date of the invitation.

sahil839 commented on zerver/worker/queue_processors.py:
```diff
@@ -460,6 +460,9 @@ def consume(self, data: Mapping[str, Any]) -> None:
         )
+        if invite_expires_in_days is None:
+            # We do not queue reminder email for never expiring invitations.
+            return
```
I am not sure whether we want to queue reminder email here. Normally we send the email two days before expiry date, but there is no expiry date for this case.

alya commented:
@sahil839 please let me know when this is ready for my review -- screenshots would be great at that point. Thanks!

sahil839 commented:
@timabbott I have rebased the PR as Mateusz changes for revoking invites on deactivation has been merged.

I have noticed one issue which occurs randomly. There is a runtime error when client fetches invites from server on receiving `invites_changed` event due to confirmation object not being created. I think this happens because we defer confirmation objects creation to queue.

@alya I am not sure what exact screenshots you are looking for, but I am attaching the screenshot of UI with the new option.
Let me know if I should share some other screenshots.

![Screenshot from 2022-01-14 16-27-40](https://user-images.githubusercontent.com/35494118/149506307-c9e6361a-6573-4735-a4af-789653da28b4.png)
![Screenshot from 2022-01-14 16-27-11](https://user-images.githubusercontent.com/35494118/149506347-39d7886b-8532-446b-a12c-e5c992cce8e6.png)

alya commented:
Cool, yeah, those are the screenshots I wanted to see -- thanks!

alya commented:
I played with the PR a bit as well, and it looks good to me.

timabbott commented on confirmation/models.py:
```diff
@@ -85,7 +86,7 @@ def create_confirmation_link(
     confirmation_type: int,
     *,
-    validity_in_days: Optional[int] = None,
+    validity_in_days: Union[Optional[int], NoValueSent] = NoValueSent(),
```
This `NoValueSent` approach is great!

timabbott replied:
Though maybe `UnspecifiedValue` would be a better name for this class, if we're also using it some non-API contexts?

sahil839 replied:
Yes `UnspecifiedValue` looks better. Updated.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -7649,6 +7649,13 @@ def do_invite_users(
+    expiry_date = confirmation_obj.expiry_date
+    if expiry_date is None:
+        return "never"
+    return datetime_to_timestamp(expiry_date)
```
Do we want to return the string `never` here?  I would have thought `None/null` would be better in the API for a non-expiring link.

sahil839 replied:
Updated it to return `None`.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -7795,9 +7800,12 @@ def do_resend_user_invite_email(prereg_user: PreregistrationUser) -> int:
+    if expiry_date is None:
+        invite_expires_in_days = None
+    else:
+        invite_expires_in_days = (expiry_date - prereg_user.invited_at).days
```
Can we add a block comment noting that we're resetting the invitation to expire as long after the reminder is sent as it lasted originally?

sahil839 replied:
Done.

timabbott commented on zerver/models.py:
```diff
@@ -2145,17 +2146,26 @@ class PreregistrationUser(models.Model):
 def filter_to_valid_prereg_users(
     query: QuerySet,
-    invite_expires_in_days: Optional[int] = None,
+    invite_expires_in_days: Union[Optional[int], NoValueSent] = NoValueSent(),
```
Can you add a docstring explaining what the `invite_expires_in_days` parameter here means?  I don't think I understand that, and am having a hard time parsing this function.

sahil839 replied:
It basically helps us from preventing sending the email if `invite_expires_in_days` number of days has passed since creating the invitation. For eg - I invited a user with expiration time set to 3 days on a particualr day, but the queue processor tries to send the email after 3 days. I am not sure whether this will actually happen.

timabbott replied:
OK. I think this logic is just wrong and we should delete support for that parameter.  AFAIK the only place we call this is here:
```
        invitee = filter_to_valid_prereg_users(                                                                                                          
            PreregistrationUser.objects.filter(id=data["prereg_id"]), invite_expires_in_days                                                             
        ).first()                                                                                                                                        
        if invitee is None:                                                                                                                              
            # The invitation could have been revoked                                                                                                     
            return                                                                                                                                       
```

and as one can see, we already have the `id` for the `PreregistrationUser` object, so we should just look it up by ID and then check directly whether it's expired/revoked, and if so, remove it.

@mateuszmandera can you help confirm my understanding here?  We should fix this in a follow-up since it's a preexisting issue, but I'd like to make sure it doesn't get lost.

timabbott commented on zerver/lib/types.py:
```diff
@@ -72,3 +72,7 @@ class SAMLIdPConfigDict(TypedDict, total=False):
+
+
+class NoValueSent:
+    pass
```
This should have a docstrings explaining how it's a solution for parameters that want to distinguish an explicit `null`.

sahil839 replied:
Done.

timabbott replied:
I did a further edit pass on this.

timabbott commented on zerver/tests/test_signup.py:
```diff
@@ -2349,6 +2374,51 @@ def test_successful_get_open_invitations(self) -> None:
+                user_profile, PreregistrationUser.INVITE_AS["MEMBER"], 100
+            )
+
+        result = self.client_get("/json/invites")
```
Maybe add a comment somewhere in this test explaining that we only expect to see the `never` ones because we mocked time to have created the other links in the deep past.

sahil839 replied:
Done.

timabbott commented on static/js/invite.js:
```diff
@@ -34,7 +34,12 @@ function reset_error_messages() {
     const invite_as = Number.parseInt($("#invite_as").val(), 10);
-    const expires_in = Number.parseFloat($("#expires_in").val());
+    let expires_in = $("#expires_in").val();
+    if (expires_in !== "null") {
```
I think this check should be for the value `null`, not the string?

sahil839 replied:
Yes, this is now `null` as we have updated the value sent from API to `None` as you suggested in one of the above comments.

timabbott commented on static/js/settings_config.ts:
```diff
@@ -467,12 +467,11 @@ export const expires_in_values = {
-    //     default: false,
-    // }
+    never: {
+        value: "null",
```
I think this should be `null`, not `"null"`.

sahil839 replied:
If we keep this value as `null` and not `"null"`, then the value returned by `.val()` for `option` element will be `""`, so I think it would be better to use "null" instead of "".

timabbott replied:
Yeah, OK, that makes sense.

timabbott replied:
Added comments documenting this somewhat unusual looking code :).

timabbott commented:
I posted a batch of comments; otherwise lgtm.

sahil839 commented:
Updated the PR. Added the docstrings and comments. I am not completely sure about the wording there especially the one in `filter_to_valid_prereg_users`. And left a comment [here](https://github.com/zulip/zulip/pull/20410#discussion_r812823314).

Also, do we want to send reminder emails for the never expiring invitations?

timabbott commented:
OK merged with a bit more work on docstrings and the like. I also added the missing API changelog entry:
```
diff --git a/templates/zerver/api/changelog.md b/templates/zerver/api/changelog.md
index a070181454..c417299979 100644
--- a/templates/zerver/api/changelog.md
+++ b/templates/zerver/api/changelog.md
@@ -21,6 +21,12 @@ format used by the Zulip server that they are interacting with.
 ## Changes in Zulip 5.0
 
 
+**Feature level 117**
+
+* `POST /invites`, `POST /invites/multiuse`: Added support for passing
+  `null` as the `invite_expires_in_days` parameter to request an
+  invitation that never expires.
+
 **Feature level 116**
 
 * [`GET /server_settings`](/api/get-server-settings): Added
diff --git a/version.py b/version.py
index a0765283f9..44fdd598f8 100644
--- a/version.py
+++ b/version.py
@@ -33,7 +33,7 @@ DESKTOP_WARNING_VERSION = "5.4.3"
 # Changes should be accompanied by documentation explaining what the
 # new level means in templates/zerver/api/changelog.md, as well as
 # "**Changes**" entries in the endpoint's documentation in `zulip.yaml`.
-API_FEATURE_LEVEL = 116
+API_FEATURE_LEVEL = 117
 
 # Bump the minor PROVISION_VERSION to indicate that folks should provision
 # only when going from an old version of the code to a newer version. Bump
```

(This is how we document API changes to endpoints without API documentation yet).  @laurynmm FYI.

@alya FYI since we may be ready to update documentation related to this.

timabbott commented:
I also added 7f2aef3ab32012c67938ddaf2853b2a22b3d7381 to correct existing missing API changelog entries for those undocumented endpoints.
